### PR TITLE
1095b

### PR DESCRIPTION
**Pull request by stephte**
## Description of change

This PR sets up the API for allowing Veterans to download their 1095B tax form from the VA. It includes a sidekiq script that pulls the tax form information from the AWS S3 bucket that it will be put in (by the Enrollment team); a Form1095B model/database table for storing the information; and a controller for downloading the PDF (plus a 1095 policy to determine form access).

## Original issue(s)
department-of-veterans-affairs/va.gov-team#35824

## Things to know about this PR
The form that the Veterans will be able to download will have PII on it, such as: last 4 of ssn; address; ICN; DOB.

There are rspec tests written, and we have tested all of the features locally, ensuring everything is working properly. We also plan on testing in staging and likely writing out some more rspec tests

### CODE REVIEW

**Review by nprimak — DISMISSED** [review]

**Comment by stephte** [comment]
I put the database stuff into a separate PR: https://github.com/department-of-veterans-affairs/vets-api/pull/9457

**Comment by stephte** [comment]
disabled robocop method length linting on the generate_pdf method in Form1095B model, the PDF fill code is more than 20 lines by itself, and cannot be split up

File: app/controllers/v0/form1095_bs_controller.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    before_action :get_1095b, only: :download
+    before_action :get_available_forms, only: :available_forms
+
+    def available_forms
+      render json: { available_forms: @available_forms_arr }
+    end
+
+    def download
+      file_name = "1095B_#{download_params}.pdf"
+      send_data @form.get_pdf, filename: file_name, type: 'application/pdf', disposition: 'attachment'
```
Comment (thilton-oddball) [inline app/controllers/v0/form1095_bs_controller.rb:15]: Not sure on this but if you're sending data here does the related route need to be a post?

File: app/models/form1095_b.rb [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+  def form_data_schema
+    {
+      "type": 'object',
+      "required": %w[first_name last_name address city coverage_months country],
+      "properties": {
+        "first_name": { "type": 'string' },
+        "middle_name": { "type": 'string' },
+        "last_name": { "type": 'string' },
+        "ssn": {
```
Comment (thilton-oddball) [inline app/models/form1095_b.rb:99]: If possible maybe name this more explicitly with something like `last_4_ssn`

File: spec/factories/form1095_bs.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+FactoryBot.define do
+  factory :form1095_b do
+    veteran_icn { '3456787654324567' }
+    tax_year { 2021 }
+    form_data {
+      {
+        first_name: 'First',
+        middle_name: 'Middle',
+        last_name: 'Last',
+        ssn: '1234',
```
Comment (thilton-oddball) [inline spec/factories/form1095_bs.rb:12]: If changed as recommended in model, will also need to be changed here.

**Review by thilton-oddball — CHANGES_REQUESTED** [review]

File: app/controllers/v0/form1095_bs_controller.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
+
+module V0
+  class Form1095BsController < ApplicationController
+    before_action { authorize :form1095, :access? }
+    before_action :get_1095b, only: :download
+    before_action :get_available_forms, only: :available_forms
```
Comment (rileyanderson) [inline app/controllers/v0/form1095_bs_controller.rb:7]: Could you use the rails standard naming convention here? `before_action :set_<instance_variable_name>`
```suggestion
    before_action :set_form, only: :download
    before_action :set_available_forms, only: :available_forms
```

File: app/controllers/v0/form1095_bs_controller.rb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+    def available_forms
+      render json: { available_forms: @available_forms_arr }
```
Comment (rileyanderson) [inline app/controllers/v0/form1095_bs_controller.rb:10]: `_arr` is redundant here.
```suggestion
      render json: { available_forms: @available_forms}
```

File: app/controllers/v0/form1095_bs_controller.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+    def download
+      file_name = "1095B_#{download_params}.pdf"
+      send_data @form.get_pdf, filename: file_name, type: 'application/pdf', disposition: 'attachment'
+    end
+
+    private
+
+    def get_1095b
+      @form = Form1095B.find_by(veteran_icn: @current_user[:icn], tax_year: download_params)
```
Comment (rileyanderson) [inline app/controllers/v0/form1095_bs_controller.rb:21]:
```suggestion
      @form = Form1095B.find_by(veteran_icn: @current_user[:icn], tax_year: download_params[:tax_year])
```

File: app/controllers/v0/form1095_bs_controller.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+      if @form.blank?
+        Rails.logger.error("Form 1095-B for year #{download_params} not found", user_uuid: @current_user&.uuid)
+        raise Common::Exceptions::RecordNotFound, download_params
+      end
+    end
+
+    def get_available_forms
```
Comment (rileyanderson) [inline app/controllers/v0/form1095_bs_controller.rb:29]:
```suggestion
    def set_available_forms
```

File: app/controllers/v0/form1095_bs_controller.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+      forms = Form1095B.get_available_forms(@current_user[:icn])
+      @available_forms_arr = forms.map { |form| { year: form[0], last_updated: form[1] } }
+    end
+
+    def download_params
+      params.require(:tax_year)
```
Comment (rileyanderson) [inline app/controllers/v0/form1095_bs_controller.rb:35]: Could you use `permit` here instead? `require` is used for nested parameters. You will then need to change your instances of `download_params` to `download_params[:tax_year]`
```suggestion
      params.permit(:tax_year)
```

File: app/models/form1095_b.rb [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+  has_kms_key
+  encrypts :form_data, key: :kms_key, **lockbox_options
+
+  # validations
+  validates :veteran_icn, :tax_year,  presence: true
+  validates :veteran_icn, uniqueness: { scope: :tax_year }
+  validate :proper_form_data_schema
+
+  # scopes
+  scope :get_available_forms, lambda { |icn|
```
Comment (rileyanderson) [inline app/models/form1095_b.rb:13]: `get` is redundant here 
```suggestion
  scope :available_forms, lambda { |icn|
```

File: app/models/form1095_b.rb [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    where(veteran_icn: icn).order(tax_year: :desc).pluck(:tax_year, :updated_at)
+  }
+
+  # methods
+  def get_pdf
```
Comment (rileyanderson) [inline app/models/form1095_b.rb:18]: `get` is redundant here.
```suggestion
  def pdf
```

File: app/models/form1095_b.rb [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+  def country_and_zip
+    "#{data['country']} #{data['zip_code'] || data['foreign_zip']}"
+  end
+
+  def middle_initial
+    data['middle_name'] ? data['middle_name'][0] : ''
+  end
+
+  # rubocop:disable Metrics/MethodLength
+  def generate_pdf
```
Comment (rileyanderson) [inline app/models/form1095_b.rb:33]: This method is quite long. Could you move some of this logic to the `pdf` method? Thoughts?
Something like:

```ruby 
def pdf 
  pdftk = PdfForms.new(Settings.binaries.pdftk)

  tmp_file = Tempfile.new("1095B-#{SecureRandom.hex}.pdf")
  pdf_template = "lib/pdf_fill/forms/pdfs/f1095bs/1095b-#{tax_year}.pdf"

  unless File.exist?(pdf_template)
    Rails.logger.error "1095-B template for year #{tax_year} does not exist."
    raise "1095-B for tax year #{tax_year} not supported"
  end

  generate_pdf(pdftk, tmp_file, pdf_template)

raise PdfForms::PdftkError => e
  ...
end


private 

def generate_pdf(pdftk, tmp_file, pdf_template)
  pdftk.fill_form(pdf_template, tmp_file, { ... }, flatten: true)

  pdf = temp_file.read
  temp_file.close
  temp_file.unlink

  pdf
end
```

File: config/sidekiq_scheduler.yml [hunk 12]
```diff
@@ -271,3 +271,4 @@ TestUserDashboard::MirrorAccountAvailabilityLogsInBigQuery:
   cron: "0 */2 * * * America/New_York"
   class: TestUserDashboard::MirrorAccountAvailabilityLogsInBigQuery
   description: "Mirrors the test_user_dashboard_tud_account_availability_logs table in BigQuery."
+
```
Comment (rileyanderson) [inline config/sidekiq_scheduler.yml:274]: Unnecessary change to this file.

File: app/models/form1095_b.rb [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_04[0]": data['coverage_months'][3] && 1,
+        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_05[0]": data['coverage_months'][4] && 1,
+        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_06[0]": data['coverage_months'][5] && 1,
+        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_07[0]": data['coverage_months'][6] && 1,
+        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_08[0]": data['coverage_months'][7] && 1,
+        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_09[0]": data['coverage_months'][8] && 1,
+        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_10[0]": data['coverage_months'][9] && 1,
+        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_11[0]": data['coverage_months'][10] && 1,
+        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_12[0]": data['coverage_months'][11] && 1,
+        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_13[0]": data['coverage_months'][12] && 1
```
Comment (rileyanderson) [inline app/models/form1095_b.rb:75]: What are you trying to do here? Are you trying to convert a boolean value to an integer? `true=1`, `false=0` ? If the data value is false `false && 1` will return `false`. If you want false to return 0 you will need to do `... && 1 || 0`. Can you add some specs to test when there are instances of false?
```suggestion
        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_01[0]": data['coverage_months'][0] && 1 || 0,
        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_02[0]": data['coverage_months'][1] && 1 || 0,
        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_03[0]": data['coverage_months'][2] && 1 || 0,
        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_04[0]": data['coverage_months'][3] && 1 || 0,
        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_05[0]": data['coverage_months'][4] && 1 || 0,
        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_06[0]": data['coverage_months'][5] && 1 || 0,
        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_07[0]": data['coverage_months'][6] && 1 || 0,
        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_08[0]": data['coverage_months'][7] && 1 || 0,
        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_09[0]": data['coverage_months'][8] && 1 || 0,
        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_10[0]": data['coverage_months'][9] && 1 || 0,
        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_11[0]": data['coverage_months'][10] && 1 || 0,
        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_12[0]": data['coverage_months'][11] && 1 || 0,
        "topmostSubform[0].Page1[0].Table1_Part4[0].Row23[0].c1_13[0]": data['coverage_months'][12] && 1 || 0
```

File: app/models/form1095_b.rb [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+    unless File.exist?(pdf_template)
+      Rails.logger.error "1095-B template for year #{tax_year} does not exist."
+      raise "1095-B for tax year #{tax_year} not supported"
+    end
+
+    pdf.fill_form(
+      pdf_template,
+      temp_file,
+      {
+        "topmostSubform[0].Page1[0].Pg1Header[0].cb_1[1]": data['is_corrected'] && 2,
```
Comment (rileyanderson) [inline app/models/form1095_b.rb:48]: What are you trying to do here? `true && 2` will return 2 and `false && 2` will return false. Is this the expected behavior?

File: app/models/form1095_b.rb [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+    )
+    pdf = temp_file.read
+
+    temp_file.close
+    temp_file.unlink
+
+    pdf
+  rescue PdfForms::PdftkError => e
+    # in case theres other errors generating the PDF
+    Rails.logger.error e
```
Comment (rileyanderson) [inline app/models/form1095_b.rb:87]:
```suggestion
    Rails.logger.error e.message
```

File: app/models/form1095_b.rb [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+    raise e
```
Comment (rileyanderson) [inline app/models/form1095_b.rb:88]: Are you trying to re-raise the error here? If so you will just need to do `raise`
```suggestion
    raise
```

**Review by rileyanderson — CHANGES_REQUESTED** [review]

File: app/controllers/v0/form1095_bs_controller.rb [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+    before_action { authorize :form1095, :access? }
+    before_action :get_1095b, only: :download
+    before_action :get_available_forms, only: :available_forms
+
+    def available_forms
+      render json: { available_forms: @available_forms_arr }
+    end
+
+    def download
+      file_name = "1095B_#{download_params}.pdf"
```
[the remark this answers is not in the document]
Comment (stephte) [reply app/controllers/v0/form1095_bs_controller.rb:14]: I guess when only having one required param, the param becomes that value (instead of a hash). Using download_params[:tax_year] errors out, because download_params only has the tax year in it (i.e. download_params == 2021)

[the remark this answers is not in the document]
Comment (stephte) [reply app/controllers/v0/form1095_bs_controller.rb:14]: wait disregard, found your .permit comment below finally

File: app/controllers/v0/form1095_bs_controller.rb (hunk 8 above)
Comment (stephte) [reply app/controllers/v0/form1095_bs_controller.rb:35]: yes could use permit, just thought that since tax_year is needed for the endpoint to work, require made sense to me. Will work either way

File: app/controllers/v0/form1095_bs_controller.rb (hunk 1 above)
Comment (stephte) [reply app/controllers/v0/form1095_bs_controller.rb:15]: GET request should work here since the endpoint is only returning something. No data is changed/saved in the API with this endpoint

File: app/models/form1095_b.rb (hunk 13 above)
Comment (stephte) [reply app/models/form1095_b.rb:75]: This is due to pdf_fill being very strange. It doesn't like the boolean value of true for checking the box in the pdf, it likes 1. A false value doesn't check the box, which is what we want. So logically here, 'boolean && 1' serves its purpose, which is to have false if false, or 1 if true

File: app/models/form1095_b.rb (hunk 14 above)
Comment (stephte) [reply app/models/form1095_b.rb:48]: I could not tell you why pdf_fill requires the number 2 in order to check this box on the pdf, but after a while of trying to figure it out, 2 worked. So yes, this is the expected behavior

File: app/models/form1095_b.rb (hunk 11 above)
Comment (stephte) [reply app/models/form1095_b.rb:33]: Yeah i like that

File: config/sidekiq_scheduler.yml (hunk 12 above)
Comment (stephte) [reply config/sidekiq_scheduler.yml:274]: this change was committed by @thilton-oddball

Comment (thilton-oddball) [reply config/sidekiq_scheduler.yml:274]: The newline at the end of the file had been removed and I put it back. This one is fine.

**Review by thilton-oddball — APPROVED** [review]

**Review by rileyanderson — APPROVED** [review]
🚀
